Sources/KSCrashRecording/KSCrashAppStateTracker.m

```diff
@@ -233,20 +249,62 @@ - (KSCrashAppTransitionState)transitionState
     return ret;
 }
 
+- (void)_locked_completeState:(KSCrashAppTransitionState)transitionState
+{
+    if (_transitionState != transitionState || _transitionComplete) {
+        return;
+    }
+    
+    _transitionComplete = YES;
+    NSLog(@"[AC] %s, completed: %d", ksapp_transition_state_to_string(transitionState), _transitionComplete);
+    
+    // send delegate
+    // ie: didCompleteTransition or something...
+}
+
+- (void)_completeState:(KSCrashAppTransitionState)transitionState
+{
+    os_unfair_lock_lock(&_lock);
+    [self _locked_completeState:transitionState];
+    os_unfair_lock_unlock(&_lock);
+}
+
 - (void)_setTransitionState:(KSCrashAppTransitionState)transitionState
 {
+    BOOL completeOnNextLoop = NO;
     NSArray<id<KSCrashAppStateTrackerObserving>> *observers = nil;
     {
         os_unfair_lock_lock(&_lock);
+        
         if (_transitionState != transitionState) {
+            
+            // finalize the current state
+            [self _locked_completeState:_transitionState];
+            
+            // move to the next state
             _transitionState = transitionState;
+            _transitionComplete = NO;
+            completeOnNextLoop = YES;
             observers = [_observers copy];
         }
+        
         os_unfair_lock_unlock(&_lock);
     }
     
-    for (id<KSCrashAppStateTrackerObserving> obs in observers) {
-        [obs appStateTracker:self didTransitionToState:transitionState];
+    if (observers) {
+        NSLog(@"[AC] %s, completed: %d", ksapp_transition_state_to_string(transitionState), _transitionComplete);
+        
+        for (id<KSCrashAppStateTrackerObserving> obs in observers) {
+            [obs appStateTracker:self didTransitionToState:transitionState];
+        }
+    }
+    
+    
+    if (completeOnNextLoop) {
+        __weak typeof(self)weakMe = self;
+        CFRunLoopPerformBlock(CFRunLoopGetCurrent(), kCFRunLoopCommonModes, ^{
+            [weakMe _locked_completeState:transitionState];
```

Comment:
So you ensure that all current processing related to the state change is fully completed before marking the transition as complete, but notify observers right away?